File: src/test/java/ev/projects/CenterTest.java
package ev.projects;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class CenterTest {

    @Test
    void playerTakingFromCenterFirstWithEmptyFloorGetsPenalty() {
        Center center = new Center();
        Floor floor = new Floor();
        Player player = new PlayerMother().newPlayer(floor);
        Game game = new Game(List.of(player, new PlayerMother().newPlayer()), center, 1);
        game.start();
        game.executeFactoryOfferPhaseWithFactory(0, game.factoryDisplays()[0].tiles()[0], 0, 4);

        player.takeTilesFromCenter(center, game.peekCenter().get(0), 0, 4);

        assertEquals(-1, floor.score());
    }

    @Test
    void playerTakingFromCenterFirstWithNonEmptyFloorGetsPenalty() {
        Center center = new Center();
        Floor floor = new Floor();
        Player player = new PlayerMother().newPlayer(floor);
        Game game = new Game(List.of(player, new PlayerMother().newPlayer()), center, 1);
        game.start();
        game.executeFactoryOfferPhaseWithFactory(0, game.factoryDisplays()[0].tiles()[0], 0, 4);

        player.takeTilesFromCenter(center, game.peekCenter().get(0), 1, 4);

        assertEquals(-2, floor.score());
    }

    @Test
    void playerTakingFromCenterSecondDoesNotGetPenalty() {
        Center center = new Center();
        Floor floor1 = new Floor();
        Player player1 = new PlayerMother().newPlayer(floor1);
        Player player2 = new PlayerMother().newPlayer();
        Game game = new Game(List.of(player1, player2), center, 0);
        game.start();
        game.changeFactoryDisplay(0, Tile.RED, Tile.RED, Tile.BLUE, Tile.YELLOW);
        game.executeFactoryOfferPhaseWithFactory(0, Tile.RED, 0, 4);
        player2.takeTilesFromCenter(center, Tile.BLUE, 0, 4);

        player2.takeTilesFromCenter(center, Tile.YELLOW, 0, 3);

        assertEquals(0, floor1.score());
    }
}

File: python/test/center_test.py
import unittest
from flask import Flask
from azul_ai_gym.game import Game
from azul_ai_gym.game_controller import GameController
from azul_ai_gym.floor import Floor
from azul_ai_gym.center import Center
from azul_ai_gym.factory_taking_request import FactoryTakingRequest
from azul_ai_gym.center_taking_request import CenterTakingRequest
from azul_ai_gym.tile import Tile
from player_mother import PlayerMother

class TestCenter(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.app = Flask(__name__)
        cls.app_context = cls.app.app_context()
        cls.app_context.push()

    @classmethod
    def tearDownClass(cls):
        cls.app_context.pop()

    def test_player_taking_from_center_first_with_empty_floor_gets_penalty(self):
        floor = Floor()
        game = Game([PlayerMother().new_player(floor=floor), PlayerMother().new_player()], Center(), 1)
        controller = GameController(game)
        controller.take_tiles_from_factory(FactoryTakingRequest(0, game.factory_displays[0].tiles[0], 0, 4))

        controller.take_tiles_from_center(CenterTakingRequest(game.peek_center()[0], 0, 4))

        self.assertEqual(-1, floor.score())

    def test_player_taking_from_center_first_with_non_empty_floor_gets_penalty(self):
        floor = Floor()
        game = Game([PlayerMother().new_player(floor=floor), PlayerMother().new_player()], Center(), 1)
        controller = GameController(game)
        controller.take_tiles_from_factory(FactoryTakingRequest(0, game.factory_displays[0].tiles[0], 0, 4))

        controller.take_tiles_from_center(CenterTakingRequest(game.peek_center()[0], 1, 4))

        self.assertEqual(-2, floor.score())

    def test_player_taking_from_center_second_does_not_get_penalty(self):
        center = Center()
        floor1 = Floor()
        player1 = PlayerMother().new_player(floor=floor1)
        player2 = PlayerMother().new_player()
        game = Game([player1, player2], center, 0)
        game.change_factory_display(0, [Tile.RED, Tile.RED, Tile.BLUE, Tile.YELLOW])
        game.execute_factory_offer_phase_with_factory(0, Tile.RED, 0, 4)
        player2.take_tiles_from_center(center, Tile.BLUE, 0, 4)

        player2.take_tiles_from_center(center, Tile.YELLOW, 0, 3)

        self.assertEqual(0, floor1.score())